tools/c7n_azure/c7n_azure/resources/network_security_group.py

```diff
@@ -168,8 +172,70 @@ def process(self, network_security_groups, event=None):
         self.source_address = self.data.get(SOURCE, None)
         self.destination_address = self.data.get(DESTINATION, None)
 
-        nsgs = [nsg for nsg in network_security_groups if self._check_nsg(nsg)]
-        return nsgs
+        match_op = self.data.get('match-operator', 'and') == 'and' and all or any
+        matching_nsg = {}
+        for nsg in network_security_groups:
+            matching_nsg['check_nsg'] = self._check_nsg(nsg)
+            if self.data.get(CIDR):
+                matching_nsg['check_cidr'] = False
+
+                permissions_to_expand = []
+                for security_rule in nsg['properties']['securityRules']:
+                    if security_rule['properties']['direction'] == self.direction_key:
+                        permissions_to_expand.append(security_rule['properties'])
+                for perm in self.expand_permissions(permissions_to_expand):
+                    if self._process_cidrs(perm):
+                        matching_nsg['check_cidr'] = True
+            matching_nsg_values = list(filter(
+                    lambda x: x is not None, matching_nsg.values()))
+
+            if match_op == all and not matching_nsg_values:
+                continue
+
+            match = match_op(matching_nsg_values)
+            if match:
+                matched.append(nsg)
+        return matched
+
+    def expand_permissions(self, permissions):
+        for p in permissions:
+            yield dict(p)
+
+    def _process_cidr(self, cidr_key, cidr_type, range_type, perm):
+        found = None
+        access_perms = self.data.get(ACCESS)
+        if perm['access'] != access_perms:
+            return False
+
+        ip_perms = perm.get(range_type, [])
+        if not ip_perms:
+            return False
+
+        if ip_perms.lower() == "internet":
+            return False
+
+        match_range = self.data[cidr_key]
+
+        if isinstance(match_range, dict):
+            match_range['key'] = cidr_type
+        else:
+            match_range = {cidr_type: match_range}
+
+        vf = ValueFilter(match_range, self.manager)
+        vf.annotate = False
+
+        found = vf({cidr_type: ip_perms})
+        return found
+
+    def _process_cidrs(self, perm):
+        found_v4 = False
+        if 'Cidr' in self.data:
+            found_v4 = self._process_cidr(
+                'Cidr',
+                'CidrIp',
+                f"{self.data['Cidr']['ipType'].lower()}AddressPrefix",
```

Comment:
Thank you for pointing this out!
I was wondering if I should change the type of `source` or `destination` fields, that are strings now, as this would be a breaking change. Another solution could be to have source (and destination) being one of `string`, `Cidr`. What do you think?

Reply:
That's a good question, I'm not sure! If we let `source` and `destination` be "string or value filter" (similar to what the AWS ingress filter does [here](https://github.com/cloud-custodian/cloud-custodian/blob/0714e36dfe2d7615d829f3dda3835f6531ab4215/c7n/resources/vpc.py#L1593-L1598)) it would let existing policies continue to work while also opening up a possibility like this:

```yaml
policies:
  - name: get-nsg-by-cidr
    resource: azure.networksecuritygroup
    filters:
        - type: ingress
          match: all
          source:
              type: value
              value_type: cidr
              op: in
              value: "10.192.0.0/20"
          ports: '443'
          ipProtocol: 'TCP'
          access: 'Allow'
```

Which seems similar to your second suggestion. I'll likely bring this up in tomorrow's community meeting though - other maintainers probably have more/better/different ideas. I don't want to derail you with one person's silliness :grin: .

Reply:
Great, keep me posted!